### PR TITLE
Changes to make unsaved kusto query work

### PR DESCRIPTION
**Pull request by gupta1**
This PR fixes #
- Makes Kusto query work for unsaved query editors
- Includes changes to make code generic at few places (provider agnostic), so that we can remove hard coded 'sql' strings at multiple places

### CODE REVIEW

File: src/sql/workbench/services/connection/browser/connectionManagementService.ts [hunk 1]
```diff
@@ -824,7 +824,7 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 			this._onConnectRequestSent.fire();
 
 			// TODO make this generic enough to handle non-SQL languages too
-			this.doChangeLanguageFlavor(uri, 'sql', connection.providerName);
+			this.doChangeLanguageFlavor(uri, this._resourceProviderService._providers[connection.providerName].providerLanguageId, connection.providerName);
```
Comment (qifahs) [inline src/sql/workbench/services/connection/browser/connectionManagementService.ts:827]: There are some extensions for PostgreSQL and MySQL Db which I believe use the MSSQL as the provider, however the connection logic is different. We do not want to break those extensions, as they will not be updated with a languageId. It might make sense to write a function here like you had. Where we check if languageId is defined and use it, if not, we hard code a mapping where for 'MSSQL' we return 'sql'.

File: src/sql/workbench/services/queryEditor/browser/queryEditorService.ts [hunk 2]
```diff
@@ -47,7 +49,7 @@ export class QueryEditorService implements IQueryEditorService {
 				let docUri: URI = URI.from({ scheme: Schemas.untitled, path: filePath });
 
 				// Create a sql document pane with accoutrements
-				const fileInput = this._editorService.createInput({ forceUntitled: true, resource: docUri, mode: 'sql' }) as UntitledTextEditorInput;
+				const fileInput = this._editorService.createInput({ forceUntitled: true, resource: docUri, mode: this._resourceProviderService._providers[connectionProviderName].providerLanguageId }) as UntitledTextEditorInput;
```
Comment (qifahs) [inline src/sql/workbench/services/queryEditor/browser/queryEditorService.ts:52]: Same comment here. We cannot rely on shipped extensions to add languageId support. Map 'MSSQL' to 'sql', if languageId is undefined.

File: src/sql/workbench/services/connection/browser/connectionManagementService.ts (hunk 1 above)
Comment (gupta1) [reply src/sql/workbench/services/connection/browser/connectionManagementService.ts]: Got it. I have added a function to handle it with few explanatory comments.

File: src/sql/azdata.d.ts [hunk 3]
```diff
@@ -2319,6 +2319,11 @@ declare module 'azdata' {
 		 */
 		displayName: string;
 
+		/**
+		 * Display language id of provider
+		 */
+		languageId: string;
```
Comment (aaomidi) [inline src/sql/azdata.d.ts:2325]: This should probably go to azdata proposed

File: src/sql/workbench/contrib/query/browser/queryInputFactory.ts [hunk 4]
```diff
@@ -26,7 +26,7 @@ const editorInputFactoryRegistry = Registry.as<IEditorInputFactoryRegistry>(Edit
 
 export class QueryEditorLanguageAssociation implements ILanguageAssociation {
 	static readonly isDefault = true;
-	static readonly languages = ['sql', 'kusto'];
+	static readonly languages = ['sql', 'kusto'];		//TODOKusto Add language id here for new languages supported in query editor. Make it easier if possible.
```
Comment (aaomidi) [inline src/sql/workbench/contrib/query/browser/queryInputFactory.ts:29]: Is this todo being tracked anywhere?

File: src/sql/workbench/services/connection/browser/connectionManagementService.ts [hunk 5]
```diff
@@ -750,6 +750,15 @@ export class ConnectionManagementService extends Disposable implements IConnecti
 		return defaultProvider && this._providers.has(defaultProvider) ? defaultProvider : undefined;
 	}
 
+	/* Gets language id of a provider if it exists */
+	private getLanguageIdIfExists(providerName: string) {
+		if (this._resourceProviderService._providers[providerName].providerLanguageId) {
```
Comment (aaomidi) [inline src/sql/workbench/services/connection/browser/connectionManagementService.ts:755]: Can be condensed to 

```
return this._resourceProviderService._providers[providerName].providerLanguageId ?? 'sql'
```

File: src/sql/workbench/services/queryEditor/browser/queryEditorService.ts [hunk 6]
```diff
@@ -99,6 +102,15 @@ export class QueryEditorService implements IQueryEditorService {
 	}
 
 	////// Private functions
+	/* Gets language id of a provider if it exists */
+	private getLanguageIdIfExists(providerName: string) {
+		if (this._resourceProviderService._providers[providerName].providerLanguageId) {
```
Comment (aaomidi) [inline src/sql/workbench/services/queryEditor/browser/queryEditorService.ts:107]: Why is this logic being repeated twice?

File: src/sql/workbench/services/resourceProvider/browser/resourceProviderService.ts [hunk 7]
```diff
@@ -17,7 +17,7 @@ import { IAdsTelemetryService } from 'sql/platform/telemetry/common/telemetry';
 export class ResourceProviderService implements IResourceProviderService {
 
 	public _serviceBrand: undefined;
-	private _providers: { [handle: string]: azdata.ResourceProvider; } = Object.create(null);
+	public _providers: { [handle: string]: azdata.ResourceProvider; } = Object.create(null);
```
Comment (aaomidi) [inline src/sql/workbench/services/resourceProvider/browser/resourceProviderService.ts:20]: This is not really recommended, use getters if you want to do this

File: extensions/mssql/src/resourceProvider/resourceProvider.ts [hunk 8]
```diff
@@ -49,11 +50,13 @@ class FireWallFeature extends SqlOpsFeature<any> {
 
 		return azdata.resources.registerResourceProvider({
 			displayName: 'Azure SQL Resource Provider', // TODO Localize
-			id: 'Microsoft.Azure.SQL.ResourceProvider',
+			id: 'MSSQL', // This should be same as the name/providerId in package.json of extension
```
Comment (aaomidi) [inline extensions/mssql/src/resourceProvider/resourceProvider.ts:53]: Is this a safe change?

File: src/sql/workbench/contrib/query/browser/queryInputFactory.ts (hunk 4 above)
Comment (gupta1) [reply src/sql/workbench/contrib/query/browser/queryInputFactory.ts:29]: @aaomidi this is a private branch we are working on currently. We will try to tackle few before checking it in to master. Depending on how many are left to be tackled, I will create tasks.

File: src/sql/azdata.d.ts (hunk 3 above)
Comment (kburtram) [reply src/sql/azdata.d.ts:2325]: yeah, all changes to `azdata.d.ts` should start in proposed and then move to the stable API once we fully are committed to maintain the change as part of the long-term extensibility API.

File: src/sql/azdata.d.ts [hunk 9]
```diff
@@ -2336,6 +2341,7 @@ declare module 'azdata' {
 	 * Represents a provider of resource
 	 */
 	export interface ResourceProvider {
+		providerLanguageId: string;
```
Comment (kburtram) [inline src/sql/azdata.d.ts:2344]: it seems unusual to have the language id communicated in this interface.  The other methods have to do with creating firewall rules.  Is this change related to firewall rules using Kusto?

**Comment by gupta1** [comment]
> I mentioned this in teams but these kind of changes / problems really need to have design review and agreement from core before the changes are made (and ideally are made by people in core who are more familiar with the code design)

As we discussed offline, I will wait for 

> I mentioned this in teams but these kind of changes / problems really need to have design review and agreement from core before the changes are made (and ideally are made by people in core who are more familiar with the code design)

**Comment by gupta1** [comment]
Closing this PR for now. As discussed offline, will wait for @anthonydresser's fixes and use that instead. For now to unblock myself, I will check in a temporary fix in Kusto feature private branch. Once the fix is ready, I will pull in those changes.
